### PR TITLE
Cleanup ImmutableArray tests

### PR DESCRIPTION
**Pull request by jamesqo**
This is still WIP, but since it's a rather large PR and almost finished I thought I'd submit it. There are still a few TODOs floating around, but feel free to review.

**Changes:**

- Theorize most of the tests to make it easier to add new test cases
- Segregate tests that assert equality from tests that check exceptions
- Standardize method/variable naming to make tests easier to read
- Fill in some gaps in the tests
  - Test more explicit interface implementations
  - Run methods that accept comparers over a variety of comparers
  - Run methods that accept collections over a variety of collections (arrays, lists, `IList<T>`, lazy enumerables, etc.)
- General cleanup
  - Replace `Assert.Equal(immutableArray1, immutableArray2)` with `Assert.True(immutableArray1 == immutableArray2)`
    - This has a poorer stack trace, but it makes it more explicit that we're testing reference equality

**Note:** I tried to keep many of the existing test cases we have, so that's why you'll see a lot of (sometimes redundant) cases in the MemberData.

/cc @AArnott @ianhays @stephentoub

### CODE REVIEW

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs [hunk 1]
```diff
@@ -790,6 +793,12 @@ public ImmutableArray<T> RemoveAt(int index)
...
         public ImmutableArray<T> RemoveRange(int index, int length)
         {
+            // TODO: There is an inconsistency here between what type of exception will be thrown
+            // if we are an IsDefault instance.
+            // If the index is less than 0, the && below will short-circuit, and self.Length will
+            // not be evaluated. Then we will get an ArgumentOutOfRangeException.
+            // Otherwise, we will get a NullReferenceException when trying to evaluate self.Length.
+
             var self = this;
             Requires.Range(index >= 0 && index <= self.Length, nameof(index));
```
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs]: There is a problem here where if `self` is default and `index` is negative, an ArgumentOutOfRangeException will be thrown. But if `index` is positive or `length` is negative, a NullReferenceException will be thrown. Would it be acceptable to add a `self.ThrowNullRefIfNotInitialized()` before this line?

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 2]
```diff
@@ -20,239 +23,326 @@ public class ImmutableArrayTest : SimpleElementImmutablesTestBase
...
+                    new StrongBox<IEnumerable<int>>(enumerable)
+                });
         }
 
-        [Fact]
-        public void CreateFromEnumerable()
+        public static IEnumerable<object[]> StrongBoxedSpecialInt32ImmutableArrayData()
         {
-            Assert.Throws<ArgumentNullException>("items", () => ImmutableArray.CreateRange((IEnumerable<int>)null));
+            // Once https://github.com/xunit/assert.xunit/pull/5 comes into corefx, all the StrongBox stuff can be removed.
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: xUnit tries to format the collection for each test case by calling `GetEnumerator` and stringifying each of the values. For default immutable arrays, this is problematic because that methods throws an InvalidOperationException. This was fixed in https://github.com/xunit/assert.xunit/pull/5, but it looks like corefx hasn't picked up that change yet. I've worked around this issue temporarily by wrapping each enumerable in a `StrongBox<T>` so it won't get formatted.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 3]
```diff
@@ -20,239 +23,326 @@ public class ImmutableArrayTest : SimpleElementImmutablesTestBase
...
-            var immutable = ImmutableArray.CreateRange(emptySource1);
-
-            // This equality check returns true if the underlying arrays are the same instance.
-            Assert.Equal(s_empty, immutable);
+            Assert.True(s_empty == source.ToImmutableArray().Clear());
         }
 
         [Fact]
-        public void CreateRangeFromImmutableArrayWithSelector()
+        public void CreateEnumerableElementType()
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Added extra tests here to ensure `Create` creates an immutable array of collections for non-arrays.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 4]
```diff
@@ -20,239 +23,326 @@ public class ImmutableArrayTest : SimpleElementImmutablesTestBase
...
+            var empty = ImmutableArray.CreateRange(new int[0]);
+            // This equality check returns true if the underlying arrays are the same instance.
+            Assert.True(s_empty == empty);
+        }
 
-            var copy1 = ImmutableArray.CreateRange(array, 0, 0, i => i + 0.5);
-            Assert.Equal(new double[] { }, copy1);
+        [Theory]
+        [MemberData(nameof(CreateRangeWithSelectorData))]
+        public void CreateRangeWithSelector<TResult>(IEnumerable<int> source, Func<int, TResult> selector, TResult dummy)
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Another xUnit issue that has been fixed, but not made it into corefx yet: the type of `TResult` cannot be inferred unless there is a `TResult` parameter that isn't wrapped. https://github.com/xunit/xunit/pull/965

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 5]
```diff
@@ -479,268 +578,513 @@ public void LastIndexOf()
...
 
-        [Fact]
-        public void ContainsEqualityComparer()
+        // TODO: This was actually calling the Enumerable extension method.
+        // There exists no such method on ImmutableArray.
+        // Also, there are no tests for a null equality comparer.
+
+        [Theory]
+        [MemberData(nameof(ContainsEqualityComparerData))]
+        public void ContainsEqualityComparer<T>(IEnumerable<T> source, T value, IEqualityComparer<T> comparer)
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: After cleaning up this test I found out this was actually calling a method in LINQ, so it isn't really relevant to ImmutableArray (aside from testing GetEnumerator, which is covered below). Should I remove this?

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 6]
```diff
@@ -888,221 +1201,431 @@ public void Remove()
...
+        public void CopyToInvalid(IEnumerable<int> source)
+        {
+            var array = source.ToImmutableArray();
+
+            // ImmutableArray<T>.CopyTo defers to Array.Copy for argument validation, so
+            // the parameter names here come from Array.Copy.
+
+            // TODO: The parameter names in CoreCLR need to be better.
+
+            Assert.Throws<ArgumentNullException>("dest", () => array.CopyTo(null));
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: The parameter names in `Array.Copy` actually match that of `ImmutableArray.CopyTo` (`sourceIndex`, `destinationIndex`, `destination`, etc.) but the names in the exceptions do not match up. I will look into changing the names in CoreCLR in a later PR.

**Comment by jamesqo** [comment]
OK, I have just finished my first pass, so un-marking this as WIP. I still have some comments/skipped tests labeled with TODO on issues I am looking for guidance on.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs [hunk 7]
```diff
@@ -728,6 +728,9 @@ public ImmutableArray<T> Replace(T oldValue, T newValue)
         public ImmutableArray<T> Replace(T oldValue, T newValue, IEqualityComparer<T> equalityComparer)
         {
             var self = this;
+            // TODO: This is calling the IImmutableList<T> overload which is silently boxing the ImmutableArray<T>.
+            // It's also causing inconsistencies where calling Replace on an IsDefault instance will throw an
+            // InvalidOperationException rather than a NullReferenceException.
```
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs]: I added some extra test coverage that called this method on a default `ImmutableArray`. That caused problems because the `IndexOf` call here is actually boxing this to an `IImmutableList<T>` and calling the [extension method](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList.cs#L171) in `ImmutableList`.

The fix would be to simply add `self.ThrowNullRefIfNotInitialized()` before this line, or modify the `IndexOf` call so it calls the `ImmutableArray` overload and avoids boxing. This would be a behavioral change from `InvalidOperationException` -> `NullReferenceException` for default arrays, but I am assuming throwing IOE in the first place because normally everything throws NRE for default arrays. Is this OK to change?

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 8]
```diff
@@ -400,62 +487,73 @@ public void CastArrayBadInterface()
...
         [Fact]
-        public void LastIndexOfDefault()
+        public void LastIndexOfInvalid()
         {
             Assert.Throws<NullReferenceException>(() => s_emptyDefault.LastIndexOf(5));
             Assert.Throws<NullReferenceException>(() => s_emptyDefault.LastIndexOf(5, 0));
             Assert.Throws<NullReferenceException>(() => s_emptyDefault.LastIndexOf(5, 0, 0));
         }
 
+        // TODO: Cleanup IndexOf/LastIndexOf tests.
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Self-note: Remove. This PR is big enough already and I can do this separately.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 9]
```diff
@@ -479,630 +577,1045 @@ public void LastIndexOf()
...
-            Assert.Throws<ArgumentOutOfRangeException>("index", () => s_empty.SetItem(0, 10));
-            Assert.Throws<NullReferenceException>(() => s_emptyDefault.SetItem(0, 10));
-            Assert.Throws<ArgumentOutOfRangeException>("index", () => s_oneElement.SetItem(1, 10));
-            Assert.Throws<ArgumentOutOfRangeException>("index", () => s_empty.SetItem(-1, 10));
+            var array = source.ToImmutableArray();
+            var expected = source.Where(i => !match(i));
+            Assert.Equal(expected, array.RemoveAll(match));
+        }
 
-            Assert.Equal(new[] { 12345 }, s_oneElemen
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Same here as with the `Contains` tests above. These are calling the Linq methods, `Concat` doesn't exist on `ImmutableArray`. Can I delete these?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs [hunk 10]
```diff
@@ -764,6 +767,7 @@ public ImmutableArray<T> Remove(T item, IEqualityComparer<T> equalityComparer)
         {
             var self = this;
             self.ThrowNullRefIfNotInitialized();
+            // TODO: This is silently boxing.
```
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs]: I discovered that there was silent boxing here too.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 11]
```diff
@@ -479,630 +577,1045 @@ public void LastIndexOf()
...
+            Assert.Throws<ArgumentOutOfRangeException>("index", () => array.RemoveAt(-1));
+            Assert.Throws<ArgumentOutOfRangeException>("length", () => array.RemoveAt(array.Length));
+            Assert.Throws<ArgumentOutOfRangeException>("index", () => array.RemoveAt(array.Length + 1));
+        }
 
-            ImmutableArray<int> removed13 = list.RemoveRange(ImmutableArray.Create(1, 3, 5));
-            Assert.Equal(1, removed13.Length);
-            Assert.Equal(new[] { 2 }, removed13);
+        [Theory]
+        [InlineData(-1, Skip = "TODO. See comment in ImmutableArray<T>.RemoveRange(int, int).")]
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: tl;dr `RemoveRange` on a default array is throwing AOORE when the index is negative, but when other parameters are invalid it's throwing NRE because it evaluates `self.Length`. Can I fix this?

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 12]
```diff
@@ -479,630 +577,1045 @@ public void LastIndexOf()
...
+        [Theory]
+        [MemberData(nameof(RemoveRangeEnumerableData))]
+        public void RemoveRangeEnumerable(IEnumerable<int> source, IEnumerable<int> items, IEqualityComparer<int> comparer)
         {
-            var updatedArray = s_manyElements.Replace(2, 10, null);
-            Assert.Equal(new[] { 1, 10, 3 }, updatedArray);
+            var array = source.ToImmutableArray();
+            IEnumerable<int> expected = items.Aggregate(
+                seed: source.ToImmutableArray(),
+                func: (a, i) => a.Remove(i, comparer));
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Self-note: This is the cleanest way to calculate `expected` but it seems uncomfortably close to the code I'm actually testing.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs (hunk 9 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: This was the one where the `IImmutableList` box was causing an IOE to be thrown for default immutable arrays.

Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: I made a review note in `SetItem`. `Requires.Range(index >= 0 && index < self.Length)`-- if index is negative the condition short-circuits.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 13]
```diff
@@ -1129,298 +1642,509 @@ public void ConcatEdgeCases()
...
+            var enumerables = Int32EnumerableData()
+                .Select(array => array[0])
+                .Cast<IEnumerable<int>>();
 
-            Assert.Equal(emptyArray.GetHashCode(EqualityComparer<int>.Default), emptyImmArray.GetHashCode(EqualityComparer<int>.Default));
-            Assert.Equal(array.GetHashCode(EqualityComparer<int>.Default), immArray.GetHashCode(EqualityComparer<int>.Default));
-            Assert.Equal(array.GetHashCode(EverythingEqual<int>.Default), immArray.GetHashCode(EverythingEqual<int>.Default));
+            return SharedComparers<int>()
+                .OfType<IEqualityComparer>()
+                .Except(new IEqualityComparer[] { null })
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Self-nit: I think all of the `Except(new[] { null })`s may not be necessary. I found out (while writing this PR) what `OfType` does not include nulls.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 14]
```diff
@@ -1129,298 +1642,509 @@ public void ConcatEdgeCases()
...
-        public void StructuralComparableArrayInterop()
+        [Theory]
+        [MemberData(nameof(IStructuralComparableCompareToData))]
+        public void IStructuralComparableCompareTo(IEnumerable<int> source, object other, IComparer comparer, int expected)
         {
-            IStructuralComparable array = new int[3] { 1, 2, 3 };
-            IStructuralComparable equalArray = new int[3] { 1, 2, 3 };
-            IStructuralComparable immArray = ImmutableArray.Create((int[])array);
-            IStructuralComparable equalImmArray = ImmutableArray.Create((int[])equalArray);
+            var array = source?.ToImmutableArray() ?? s_emptyDefault;
```
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: I got a bit lazy here and used `null` for source/other to represent a default immutable array, instead of a `StrongBox`.

**Comment by jamesqo** [comment]
Test Innerloop Windows_NT Release Build and Test

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs (hunk 7 above)
Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs]: Avoiding boxing sounds reasonable. But let's make that change in another PR and keep this one about testing.
I would suggest you revert the TODO comment and file an issue to track.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs (hunk 1 above)
Comment (AArnott) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs]: Either exception should never be caught and is a sign of an error in the caller. So consistently prioritizing one over the other sounds like (mild) goodness to me.
But unless it's blocking your test development, I suggest you fix this in a separate PR as well.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs (hunk 7 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs]: @AArnott OK, I have reverted the comment and opened https://github.com/dotnet/corefx/issues/14961 to be addresed after the test changes go in.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 15]
```diff
@@ -261,13 +350,13 @@ public void Covariance()
...
             Assert.False(baseImmutable.IsDefault);
-            // Must cast to object or the IEnumerable<object> overload of Equals would be used
+            // Must cast to object or the IEnumerable<object> overload of Assert.Equal would be used
             Assert.Equal((object)derivedImmutable, baseImmutable, EqualityComparer<object>.Default);
 
             // Make sure we can reverse that, as a means to verify the underlying array is the same instance.
             ImmutableArray<string> derivedImmutable2 = baseImmutable.As<string>();
             Assert.False(derivedImmutable2.IsDefault);
-            Assert.Equal(derivedImmutable, derivedImmutable2);
+            Assert.True(derivedImmutable == derivedImmutable2);
```
Comment (ianhays) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Why the change from Assert.Equal?

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 16]
```diff
@@ -479,948 +575,1525 @@ public void LastIndexOf()
...
 
-        [Fact]
-        public void InsertDefault()
+        public static IEnumerable<object[]> EqualsData()
         {
-            Assert.Throws<NullReferenceException>(() => s_emptyDefault.Insert(-1, 10));
-            Assert.Throws<NullReferenceException>(() => s_emptyDefault.Insert(1, 10));
-            Assert.Throws<NullReferenceException>(() => s_emptyDefault.Insert(0, 10));
-        }
+            // Once https://github.com/xunit/assert.xunit/pull/5 comes into corefx, all the StrongBox stuff can be removed.
```
Comment (ianhays) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Instead of doing this workaround, why don't we bring those necessary changes into buildtools?

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs [hunk 17]
```diff
@@ -479,948 +575,1525 @@ public void LastIndexOf()
...
+        }
+
+        [Theory]
+        [MemberData(nameof(Int32EnumerableData))]
+        public void RemoveAllInvalid(IEnumerable<int> source)
+        {
+            var array = source.ToImmutableArray();
+
+            Assert.Throws<ArgumentNullException>("match", () => array.RemoveAll(match: null));
         }
```
Comment (ianhays) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Same here, lets look into bringing this into buildtools so we don't have to work around its absence.

**Comment by ianhays** [comment]
Thanks @jamesqo, these look good. Sorry for the delay in the review, there are quite a few changes here :) Did you do a coverage comparison to make sure no coverage of value was lost?

**Comment by jamesqo** [comment]
> Did you do a coverage comparison to make sure no coverage of value was lost?

I looked into the code coverage reports and all lines in the type are still covered.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs (hunk 16 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: Nice idea. I will look into making a PR to buildtools that updates xunit.

File: src/System.Collections.Immutable/tests/ImmutableArrayTest.cs (hunk 15 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/tests/ImmutableArrayTest.cs]: @ianhays When I first saw `Assert.Equal` I was confused because I thought it was calling the `Equal(IEnumerable<T>, IEnumerable<T>)` overload which compares the contents of the arrays. Hovering over the method in VS showed that this was not the case, but I still think this is better because the tests will fail loudly if you accidentally write `Assert.True(expected == source)` (where `source` is an enumerable and `expected` is an ImmutableArray) whereas they will silently pass if you write `Assert.Equal(expected, source)`.
